src/main/ruby/core/encoding.rb

```diff
@@ -48,34 +48,34 @@ class << self
       map
     end
 
-    private def setup_default_encodings
-      # Note the locale Encoding is patched below,
-      # so modify that code too if you change the code here.
-      %w[internal external locale filesystem].each do |name|
-        key = name.upcase.to_sym
-        enc = Truffle::EncodingOperations.get_default_encoding(name)
-        index = enc ? EncodingMap[enc.name.upcase.to_sym].last : nil
-        EncodingMap[key] = [name, index]
-      end
+    private def setup_default_encoding(name, key)
+      enc = Truffle::EncodingOperations.get_default_encoding(name)
+      index = if enc
+                # UTF-8 is the default value for those, so optimize for that
+                enc_name = enc == Encoding::UTF_8 ? :'UTF-8' : enc.name.upcase.to_sym
+                EncodingMap[enc_name].last
+              else
+                nil
+              end
+      EncodingMap[key] = [name, index]
+      enc
     end
   end
 
   EncodingMap = build_encoding_map
-  setup_default_encodings
+
+  @default_internal = setup_default_encoding('internal', :INTERNAL)
```

Comment:
While a bit more verbose, I think the `EncodingMap` should be passed as an argument here. It's not really clear from this call that it must occur after the `build_encoding_map` call. Prior to this refactoring, that distinction was unnecessary. I can appreciate why you split the method up, but I think some amount of cohesion was lost in the process.